test_conformance/extensions/cl_khr_command_buffer/command_buffer_queue_substitution.cpp

```diff
@@ -0,0 +1,349 @@
+          || BasicCommandBufferTest::Skip();
+    }
+
+    //--------------------------------------------------------------------------
+    cl_int CreateCommandQueueWithProperties(cl_command_queue &queue_with_prop)
+    {
+        cl_int error = 0;
+        cl_queue_properties_khr device_props = 0;
+
+        error = clGetDeviceInfo(device, CL_DEVICE_QUEUE_PROPERTIES,
+                                sizeof(device_props), &device_props, nullptr);
+        test_error(error,
+                   "clGetDeviceInfo for CL_DEVICE_QUEUE_PROPERTIES failed");
+
+        using PropPair = std::pair<cl_queue_properties_khr, std::string>;
+
+        auto check_property = [&](const PropPair & prop)
+        {
+          if (device_props & prop.first)
+          {
+            log_info("Queue property %s supported. Testing ... \n",
+                     prop.second.c_str());
+            queue_with_prop = clCreateCommandQueue
+                (context, device, prop.first, &error);
+          }
+          else
+              log_info("Queue property %s not supported \n", prop.second.c_str());
+        };
+
+        // in case of extending property list in future
+        std::vector< PropPair > props = {
+          ADD_PROP(CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE),
+          ADD_PROP(CL_QUEUE_PROFILING_ENABLE)
+        };
+
+        for ( auto && prop : props )
+        {
+          check_property(prop);
+          test_error(error, "clCreateCommandQueue failed");
+          if (queue_with_prop!=nullptr)
+              return CL_SUCCESS;
+        }
+
+        return CL_INVALID_QUEUE_PROPERTIES;
+    }
+
+    //--------------------------------------------------------------------------
+    cl_int SetUp(int elements) override
+    {
+        // By default command queue is created without properties,
+        // if test requires queue with properties default queue must be replaced.
+        if (properties_use_requested)
+        {
+            // due to the skip condition
+            queue = nullptr;
+            if ( CreateCommandQueueWithProperties(queue) != CL_SUCCESS )
+              return CL_SUCCESS;
+        }
+
+        return BasicCommandBufferTest::SetUp(elements);
+    }
+
+    //--------------------------------------------------------------------------
+    cl_int Run() override
+    {
+        cl_int error = CL_SUCCESS;
+        cl_command_queue_properties cqp;
+        error = clGetCommandQueueInfo(queue, CL_QUEUE_PROPERTIES, sizeof(cqp), &cqp, NULL);
+        test_error(error, "clGetCommandQueueInfo failed");
+
+        if(properties_use_requested && (cqp & CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE))
+        {
+          out_of_order_queue = true;
+        }
+
+        // record command buffer with primary queue
+        error = RecordCommandBuffer();
+        test_error(error, "RecordCommandBuffer failed");
+
+        // create substitute queue
+        cl_command_queue new_queue = nullptr;
```

Comment:
Can use `clCommandQueueWrapper` here, it reference counts and automatically calls `clReleaseCommandQueue` when the last reference gets deleted.